### DISCUSSION TITLE
AI Integration

### DISCUSSION THREAD

**Discussion by antonvh** [TandoorRecipes/recipes#2747] [Ideas]
Hi all! I just installed tandoor via docker/portainer. It works really nice.

I was looking into tandoor because a friend showed me crouton app, which can import recipes from a smartphone photo. I would like something similar and it probably requires AI to identify ingredients and amounts in recipe photo and text.

So far I was able to do it with OpenAI/ChatGPT. It can read photos. It can also convert a spoken explanation of a recipe into a nicely structured and written recipe. It can even generate a photo from the written recipe.

I got stuck with Tandoor in the import section. I asked ChatGPT to format the recipe as JSON, tried pasting that in source import, but that didn't work.

I'm reasonably fluid in Python, I never worked with Django, and I know very little of the Tandoor architecture and roadmap. What would be a good way to integrate an AI API into the source?

https://platform.openai.com/docs/api-reference/making-requests

**Comment by vabene1111** [comment]
interesting, using chat GPT for that would probably work. 

Many people who self host prefer to have everything run locally on their device so the ideal solution would have some kind of pre-trained model available directly in the application that is specialised for these kinds of tasks and can run on a lighter infrastructure but I am not sure if AI is advanced enough for that.

Do you know any more about this or want to figure it out? Could you also post a few examples of image to json conversions that chat GPT can handle and which promts you used so I can get a feel of how good it is. maybe integrating with open AI as a optional component would be something useful for people.

**Comment by antonvh** [comment]
The prompt is simple English really:
<img width="989" alt="Screenshot 2023-11-27 at 21 14 06" src="https://github.com/TandoorRecipes/recipes/assets/355996/dc563c80-491d-4c14-ac13-52dd4c84c7e2">

It then generates this recipe in plain text:
Here's the step-by-step recipe for Shakshuka:

Ingredients:

2 tbsp olive oil
2 tbsp Pilpelchuma or harissa paste
2 tsp tomato purée
2 large red peppers, cut into 0.5cm dice (300g in total)
4 garlic cloves, finely chopped
1 tsp ground cumin
5 large, very ripe tomatoes, chopped (800g in total); tinned are also fine
4 medium free-range eggs plus 4 egg yolks
120g Labneh or thick yoghurt
Salt
Instructions:

1. Heat olive oil in a large frying pan over medium heat.
2. Add the Pilpelchuma or harissa, tomato purée, peppers, garlic, and cumin. Season with ¾ tsp of salt.
3. Stir and cook until the peppers soften, about 8 minutes.
4. Add the tomatoes and bring to a gentle simmer. Cook for 10 minutes until you have a thick sauce. Taste for seasoning.
5. Make eight dips in the sauce. Break the eggs and carefully pour each into its own dip. Do the same with the yolks.
6. Use a fork to swirl the egg whites a bit with the sauce, taking care not to break the yolks.
7. Simmer gently for 8-10 minutes, until the egg whites are set but the yolks are still runny. Cover with a lid to hasten the process if you wish.
8. Remove from the heat, let it settle for a couple of minutes, then serve on individual plates with Labneh or yoghurt.

And, when asked generates semantic html. But you can feed it any json format, and chatgpt will follow that example.
```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Shakshuka Recipe</title>
</head>
<body>

<article itemscope itemtype="http://schema.org/Recipe">
  <header>
    <h1 itemprop="name">Shakshuka</h1>
  </header>
  
  <img itemprop="image" src="shakshuka.jpg" alt="Delicious Shakshuka">
  
  <span itemprop="author" itemscope itemtype="http://schema.org/Person">
    Author: <span itemprop="name">Chef Name</span>
  </span>
  
  <meta itemprop="datePublished" content="2023-11-27">
  <span>Published on: November 27, 2023</span>

  <section itemprop="description">
    <p>Shakshuka is a Tunisian dish that's popular in Jerusalem and all over Israel. It's a delicious breakfast or lunch fare, with a seasonal variant that includes eggs, tomato, and spices.</p>
  </section>

  <section itemprop="nutrition" itemscope itemtype="http://schema.org/NutritionInformation">
    <h2>Nutritional Information</h2>
    <p>Calories: <span itemprop="calories">250</span></p>
    <p>Protein: <span itemprop="proteinContent">8.25g</span> per serving</p>
  </section>

  <section itemprop="recipeIngredient">
    <h2>Ingredients</h2>
    <ul>
      <li>2 tbsp olive oil</li>
      <!-- More ingredients -->
    </ul>
  </section>

  <section itemprop="recipeInstructions">
    <h2>Instructions</h2>
    <ol>
      <li itemprop="recipeStep">Heat olive oil in a large frying pan over medium heat.</li>
      <!-- More steps -->
    </ol>
  </section>
  
  <section itemprop="aggregateRating" itemscope itemtype="http://schema.org/AggregateRating">
    <h2>Rating</h2>
    <p>
      Rated: <span itemprop="ratingValue">4.5</span>/5 based on <span itemprop="reviewCount">150</span> reviews
    </p>
  </section>
  
  <section itemprop="review" itemscope itemtype="http://schema.org/Review">
    <h2>Reviews</h2>
    <article>
      <p itemprop="reviewBody">This shakshuka recipe is a game-changer for brunch! Absolutely delicious and full of flavor.</p>
      <footer>
        <p>Review by: <span itemprop="author">Jane Doe</span></p>
      </footer>
    </article>
    <!-- More reviews -->
  </section>
  
</article>

</body>
</html>
```

Fun fact: it can estimate nutritional values too:
To estimate the protein content of the Shakshuka recipe, we'll look at the protein content of the main ingredients known for their protein:

Eggs: On average, a large egg has about 6 grams of protein. Since the recipe calls for 4 whole eggs plus 4 additional yolks, the whites are the main source of protein. We'll count only the 4 whole eggs for protein estimation.
Labneh: This strained yogurt can have between 5 to 10 grams of protein per 100 grams depending on its thickness and concentration. We'll use an average value of 7.5 grams for 120 grams.
Harissa Paste: It contains a negligible amount of protein.
Adding these up:

4 eggs: 4 x 6g = 24 grams
120g labneh: 1.2 x 7.5g = 9 grams
Total protein for the recipe: 24g (eggs) + 9g (labneh) = 33 grams

Divided by 4 servings, each serving would approximately have:

33g / 4 = 8.25 grams of protein per serving.

This is a rough estimation. For precise nutritional information, you would need to consider all ingredients and their amounts, including those with smaller amounts of protein, and use nutritional data sources or software.

**Comment by antonvh** [comment]
I forced it to insert all ingredients and steps. Sadly Dall-e seems to be down, otherwise it would have added a tasty image.
```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Shakshuka Recipe</title>
</head>
<body>

<article itemscope itemtype="http://schema.org/Recipe">
  <header>
    <h1 itemprop="name">Shakshuka</h1>
    <img itemprop="image" src="shakshuka.jpg" alt="Delicious Shakshuka">
  </header>
  
  <meta itemprop="datePublished" content="2023-11-27">
  <span>Published on: November 27, 2023</span>

  <section itemprop="description">
    <p>Shakshuka is a Tunisian dish that's popular in Jerusalem and all over Israel. It's a delicious breakfast or lunch fare, with a seasonal variant that includes eggs, tomato, and spices.</p>
  </section>

  <section itemprop="nutrition" itemscope itemtype="http://schema.org/NutritionInformation">
    <h2>Nutritional Information</h2>
    <p>Calories: <span itemprop="calories">Estimated 250</span></p>
    <p>Protein: <span itemprop="proteinContent">Estimated 8.25g</span> per serving</p>
    <!-- Include other nutritional information if known -->
  </section>

  <section itemprop="recipeIngredient">
    <h2>Ingredients</h2>
    <ul>
      <li>2 tbsp olive oil</li>
      <li>2 tbsp Pilpelchuma or harissa paste (see page 302 or use shop-bought)</li>
      <li>2 tsp tomato purée</li>
      <li>2 large red peppers, cut into 0.5cm dice (300g in total)</li>
      <li>4 garlic cloves, finely chopped</li>
      <li>1 tsp ground cumin</li>
      <li>5 large, very ripe tomatoes, chopped (800g in total); tinned are also fine</li>
      <li>4 medium free-range eggs plus 4 egg yolks</li>
      <li>120g Labneh (see recipe page 302) or shop-bought, or thick yoghurt</li>
      <li>Salt to taste</li>
    </ul>
  </section>

  <section itemprop="recipeInstructions">
    <h2>Instructions</h2>
    <ol>
      <li itemprop="recipeStep">Heat the olive oil in a large frying pan over medium heat.</li>
      <li itemprop="recipeStep">Add the Pilpelchuma or harissa, tomato purée, peppers, garlic, cumin and ¾ of a teaspoon of salt. Stir and cook on a moderate heat for about 8 minutes to allow the peppers to soften.</li>
      <li itemprop="recipeStep">Add the tomatoes, bring to a gentle simmer and cook for a further 10 minutes until you have quite a thick sauce. Taste for seasoning.</li>
      <li itemprop="recipeStep">Make eight little dips in the sauce. Gently break the eggs and carefully pour each into its own dip. Do the same with the yolks. Use a fork to swirl the egg whites a little bit with the sauce, taking care not to break the yolks.</li>
      <li itemprop="recipeStep">Simmer gently for 8–10 minutes, until the egg whites are set but the yolks are still runny (you can cover the pan with a lid if you wish to hasten the process).</li>
      <li itemprop="recipeStep">Remove from the heat, let it settle for a couple of minutes, then spoon into individual plates and serve with the labneh or yoghurt.</li>
    </ol>
  </section>
  
  <footer>
    <section itemprop="aggregateRating" itemscope itemtype="http://schema.org/AggregateRating">
      <h2>Rating</h2>
      <p>
        Rated: <span itemprop="ratingValue">4.5</span>/5 based on <span itemprop="reviewCount">150</span> reviews
      </p>
    </section>
    
    <section itemprop="review" itemscope itemtype="http://schema.org/Review">
      <h2>Reviews</h2>
      <article>
        <p itemprop="reviewBody">This shakshuka recipe is a game-changer for brunch! Absolutely delicious and full of flavor.</p>
        <footer>
          <p>Review by: <span itemprop="author">Jane Doe</span></p>
        </footer>
      </article>
      <!-- More reviews -->
    </section>
  </footer>

</article
```

**Comment by vabene1111** [comment]
that looks really cool, I have no idea when I will be able to work on this but if anyone wants to pick it up I suggest to implement it in a way that a user can add an API token to OpenAI (or oauth or something) and then add a special page to the ImportView that allows images to be send to OpenAI. The response can either be directly or parsed into the format returned by recipe-scrapers or a special scraper could be build to then push the recipe into the normal parsing flow as any normally URL imported recipe is.

**Comment by lagset** [comment]
Open AIs ChatGPT integration would be helpful for my use case as well.

I am not using images, but recipes published on Instagram (please don't judge me 😅) in instructered plain text. And I want these out of the walled garden in a more usable interface.

I had success by manually using ChatGPT 3.5 to convert the plain text into JSON and then importing the generated JSON to tandoor. If this workflow could be automated (ideally, provide link or plain text) this would be great.

**Reply by BobWs** [reply]
Any progress on this one year later?

**Reply by antonvh** [reply]
I'm using n8n for this now. I would prefer tandoor to focus on the API, the JSON and the UI developments.

**Reply by BobWs** [reply]
> I'm using n8n for this now. I would prefer tandoor to focus on the API, the JSON and the UI developments.

Don’t know n8n, but I tried this a couple of months ago and I was surprised how accurate GPT could read my scanned handwritten family recipes! [#3585](https://github.com/TandoorRecipes/recipes/discussions/3585)
But as always with online AI use I’m careful with posting information about anything personal, so also with these handwritten family recipes! If I can manage to get this working with offline LLM then it would be great.
Unfortunate I haven’t been able to do so.

**Reply by vabene1111** [reply]
Tandoor 2 has an AI integration that can parse pictures and handwritten PDF's. The results I got when testing with gemini 2.0 flash were really good. You can choose the provider/model trough configuration but I have only tested gemini.

**Reply by antonvh** [reply]
Great work! I'm unsure how to install the alpha with my current docker setup, so I'll wait for the release.

**Reply by BobWs** [reply]
> You can choose the provider/model trough configuration but I have only tested gemini.

Will this also work with offline models like [gpt4all](https://home.nomic.ai/gpt4all)?
Would be great to use AI offline when one concerns about privacy!

**Reply by BobWs** [reply]
Where can I find the alpha image?

**Reply by lagset** [reply]
> You can install it by switching your image to vabene1111/recipes:feature-vue3/ghcr.io/tandoorrecipes/recipes:feature-vue3 to always get the latest version or to the tagged released for the alpha versions. Please keep in mind that things can and will break during the alpha so you should probably use the tagged releases and not auto update.

See: https://github.com/TandoorRecipes/recipes/releases

**Reply by vabene1111** [reply]
If you provide some kind of API endpoint you can use private AI servers already, I use a library under the hood that supports multiple different providers. 

Since this is relatively new to me and the landscape of AI is changing so fast we will need to work on providing proper documentation and maybe some adapters to change the data formats/messaging depending on the platform that is used. 

If anyone is really into this feel free to contact me, maybe we can choose a good architecture for this for the long term future.

**Reply by ClemensSchartmueller** [reply]
good to see some progress! I had some success with using gemini to extract recipes from screenshots of other apps and export them to the tandoor schema. I recently also added rudimentary direct API integration to it too: https://github.com/ClemensSchartmueller/AIRecipeExtractor. Just be aware that this is a vibe coding experiment so use at your own risks. It may be worth to play around with until Tandoor 2.0 releases, though :)

**Comment by johnmacdowall** [comment]
I'm doing something similar!  I already have the recipes in a MySQL database.  I have the code to extract the recipes, but I'm having a horrible time trying to post the data to Tandoor using the API.  I can create a recipe, add the steps of directions, but then when I try to add the ingredients, through a find or create routine, I start getting complaints.  When I look at the recipe with my browser, I see the recipe, there's steps but no ingredients.  When I try to edit, there's no place to add ingredients!  It seems like I have to add the ingredients to the data base and then link then back to the recipe.  This is hard!  LOL

**Comment by hhopke** [comment]
Looking really forward to see here some progress. I actually wanted to checkout tandoor 2.0 implementation but docs in todays release notes were not accessible. FYI @vabene1111

**Reply by vabene1111** [reply]
here is the link to the docs https://docs.tandoor.dev/system/configuration/#ai-integration 

I will check the release notes, thanks for the heads up

**Comment by anytimesoon** [comment]
Has anyone managed to get the ai feature working with a self hosted ai? The docs seem to imply that it should be possible but it's not clear how that could be done

**Reply by vabene1111** [reply]
just enter the API key for a google API account and it should work. If you change the models it might as well but I have not tested that

**Reply by anytimesoon** [reply]
So it only works with google ai at the moment?

What is meant by the AI provider in the documentation? 

`To use AI to perform different tasks you need to configure an API key and the AI provider.`

**Comment by hhopke** [comment]
@vabene1111 would it be possible to do separation into steps (Schritte) automatically by AI import? I just wonder how hard it would be. I do have some basic python programming skills maybe I could attempt.

This way it would be easy to follow stepwise introductions provided. ATM its imported as one.

**Reply by vabene1111** [reply]
I plan on adding multiple AI endpoints over the time. One of them is to have an AI automatically split stuff into steps. 

Also the importer in general needs to be changed to support steps when the source supports them. When the code was started recipe-scrapers did not support this very well but now it does so the import process can be improved

**Comment by jstrvr** [comment]
Maybe helpful to share how I currently use AI and tandoor. I have a space (similar to a gpt) in perplexity with these settings and a recipeformat attached. I can include multiple images or just a link, and it converts it. I import this with the JSON/HTML option. Any improvements or ideas are welcome. I like the idea of adding a list of existing foods in my database to this space. Additionally, I use the kitshn app to import recipes from instagram.

You are tasked with converting a recipe from a given source (which could be a picture, link, or video) into a standardized HTML format. Your goal is to accurately extract the recipe information and present it in the provided HTML template.

For every question, first analyze the content of 'recipeformat.txt'. as the main source for the html structure.

### Custom prompt

Now, carefully analyze the recipe source provided:

<recipe_source>
{{RECIPE_SOURCE}}
</recipe_source>

Follow these steps to extract the recipe:

1. Identify the recipe title.
2. Extract the list of ingredients.
3. Determine the number of servings (pick the lowest when a range is given).
4. Identify the preparation time and cooking time.
5. Extract the preparation steps.
6. If available, identify any additional notes or tips.
7. I available, identify the name of the chef and add this as a keyword.
8. Depending on the estimated nutritional value add keyword: 'healthy', 'balanced' or 'cheatmeal'.
9. Depending on required cooking time add keyword: 'weekday' (for quick recipes) or 'weekend' (extensive recipes).
10. Depending on the type of recipe add keyword: 'Starter', 'Salad','Dinner','Dessert','Pickle' or 'Snack'.
11. Depending on the recipe add the keyword:’Spring’,’Summer’,’Autumn’,’Winter’.
12. Provide a descriptive name for every recipe step.

When filling out the template, make sure to:
- Preserve the exact wording from the original recipe as much as possible.
- Make sure all quantities use the metric system.
- Make sure all ingredients follow the following structure [amount(number only)  unit ingredient, notes].
- Use 'recipeformat.txt' entities for special characters (e.g., &deg; for degree symbol).
- If the recipe comes from a link also include the images from that link.
- Ensure all measurements are consistent (e.g., convert all measurements to the same unit system if necessary).
- If any information is not available in the recipe source, leave the corresponding HTML tag empty.
- Always add the keyword "Try" to a new recipe.

Your final output should be the completed HTML template with all the recipe information filled in. Do not include any explanations or additional text outside of the HTML structure. Always use 'recipeformat.txt' as the main source for the html structure.

### Recipeformat.txt

```
<!DOCTYPE html>
<html lang="en">
<head>
  <title>Sample Recipe</title>
  <script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "Recipe",
    "name": "Mediterranean Stuffed Zucchini",
    "description": "Mediterranean Stuffed Zucchini topped with breadcrumbs and cheese.",
    "image": "zucchini.jpg",
    "keywords": "Thai, dessert, mango, sticky rice, coconut, vegan, gluten-free, summer, Asian sweets",
    "recipeIngredient": [
      "Zucchini",
      "Olive oil",
      "Onion",
      "Garlic",
      "Ground lamb"
    ],
    "recipeInstructions": [
      {
        "@type": "HowToStep",
        "text": "Preheat oven to 450 degrees F (230 degrees C)."
      },
      {
        "@type": "HowToStep",
        "text": "Hollow out zucchini and prepare filling."
      }
    ]
  }
  </script>
</head>
</html>
```

Output

```<!DOCTYPE html> <html lang="en"> <head> <title>Smashed Potato Salad</title> <script type="application/ld+json"> { "@context": "https://schema.org", "@type": "Recipe", "name": "Smashed Potato Salad", "description": "Smashed Potato Salad with crispy baby potatoes, fresh cucumber, herbs, and a creamy soy yogurt dressing.", "image": "", "keywords": "Try, Salad, Dinner, Summer, balanced, vegan, healthy, weekday, potato, gluten-free, Pick Up Limes, Sadia", "recipeYield": "4", "prepTime": "PT15M", "cookTime": "PT1H10M", "totalTime": "PT1H25M", "recipeIngredient": [ "1000 g baby potatoes", "60 ml olive oil", "2 g ground cumin", "2.3 g onion powder", "0.5 g sweet paprika powder", "0.5 g salt", "195 g unsweetened soy yogurt", "120 g vegan mayonnaise", "1 lemon, juiced", "1 garlic clove, crushed", "10 ml dijon mustard", "8 ml agave syrup", "0.25 g salt", "1 cucumber, chopped", "0.5 small red onion, minced", "2 g fresh parsley, chopped", "2 g fresh dill, chopped" ], "recipeInstructions": [ { "@type": "HowToStep", "name": "Boil the potatoes", "text": "Cook the potatoes in a large pot of salted simmering water for 15 - 20 minutes until tender." }, { "@type": "HowToStep", "name": "Prepare spiced oil", "text": "Meanwhile, mix the olive oil, ground cumin, onion powder, sweet paprika powder, and salt in a small bowl. Set aside." }, { "@type": "HowToStep", "name": "Preheat oven and flatten potatoes", "text": "Drain and pat dry the potatoes. Line two baking trays and preheat the oven to 220&deg;C (425&deg;F). Flatten the potatoes on the baking tray into 1 cm thick disks using the bottom of a cup." }, { "@type": "HowToStep", "name": "Bake potatoes first round", "text": "Brush the spiced oil onto the smashed potatoes, reserving leftover oil. Bake for 20 minutes." }, { "@type": "HowToStep", "name": "Bake potatoes second round", "text": "Flip the potatoes, brush on leftover oil, and bake for another 20 - 30 minutes until golden and crispy." }, { "@type": "HowToStep", "name": "Prepare the dressing", "text": "Meanwhile, mix together soy yogurt, vegan mayonnaise, lemon juice, crushed garlic, dijon mustard, agave syrup, and salt in a large bowl." }, { "@type": "HowToStep", "name": "Combine veggies", "text": "Add chopped cucumber, minced red onion, fresh parsley, and dill to the dressing. Mix well." }, { "@type": "HowToStep", "name": "Assemble and serve", "text": "Let the baked potatoes cool for a few minutes, toss them with the veggies and dressing, and serve." } ], "servingSize": "4", "nutrition": "", "aggregateRating": "", "author": { "@type": "Person", "name": "Sadia" } } </script> </head> <body> <h1>Smashed Potato Salad</h1> <p>Smashed Potato Salad with crispy baby potatoes, fresh cucumber, herbs, and a creamy soy yogurt dressing.</p> <img src="" alt="Smashed Potato Salad"> <h2>Keywords</h2> <p>Try, Salad, Dinner, Summer, balanced, vegan, healthy, weekday, potato, gluten-free, Pick Up Limes, Sadia</p> <h2>Ingredients (metric)</h2> <ul> <li>1000 g baby potatoes</li> <li>60 ml olive oil</li> <li>2 g ground cumin</li> <li>2.3 g onion powder</li> <li>0.5 g sweet paprika powder</li> <li>0.5 g salt</li> <li>195 g unsweetened soy yogurt</li> <li>120 g vegan mayonnaise</li> <li>1 lemon, juiced</li> <li>1 garlic clove, crushed</li> <li>10 ml dijon mustard</li> <li>8 ml agave syrup</li> <li>0.25 g salt</li> <li>1 cucumber, chopped</li> <li>0.5 small red onion, minced</li> <li>2 g fresh parsley, chopped</li> <li>2 g fresh dill, chopped</li> </ul> <h2>Servings</h2> <p>4</p> <h2>Preparation Time</h2> <p>15 min</p> <h2>Cooking Time</h2> <p>1 hr 10 min</p> <h2>Steps</h2> <ol> <li><b>Boil the potatoes:</b> Cook the potatoes in a large pot of salted simmering water for 15 - 20 minutes until tender.</li> <li><b>Prepare spiced oil:</b> Meanwhile, mix the olive oil, ground cumin, onion powder, sweet paprika powder, and salt in a small bowl. Set aside.</li> <li><b>Preheat oven and flatten potatoes:</b> Drain and pat dry the potatoes. Line two baking trays and preheat the oven to 220&deg;C (425&deg;F). Flatten the potatoes on the baking tray into 1 cm thick disks using the bottom of a cup.</li> <li><b>Bake potatoes first round:</b> Brush the spiced oil onto the smashed potatoes, reserving leftover oil. Bake for 20 minutes.</li> <li><b>Bake potatoes second round:</b> Flip the potatoes, brush on leftover oil, and bake for another 20 - 30 minutes until golden and crispy.</li> <li><b>Prepare the dressing:</b> Meanwhile, mix together soy yogurt, vegan mayonnaise, lemon juice, crushed garlic, dijon mustard, agave syrup, and salt in a large bowl.</li> <li><b>Combine veggies:</b> Add chopped cucumber, minced red onion, fresh parsley, and dill to the dressing. Mix well.</li> <li><b>Assemble and serve:</b> Let the baked potatoes cool for a few minutes, toss them with the veggies and dressing, and serve.</li> </ol> <h2>Additional Notes</h2> <ul> <li>We used top and bottom heating (not fan-assisted/convection).</li> <li>This salad is best when enjoyed immediately after assembling.</li> <li>Alternatively, store the baked smashed potatoes and salad dressing ingredients separately in airtight containers in the fridge for up to 2 days.</li> </ul> </body> </html>```

Another case I use AI for is converting a menu. I then give perplexity this prompt: create 1 recipe where each dish is a separate step with ingredients

**Reply by juantxorena** [reply]
To be honest, this should be part of the application. Right now there is a way to import URLs of recipes, and a way to use AI to import images. But the obvious usecase (in my opinion) of importing recipes from URLs, and using AI format them properly and translate them to another language if needed is missing.

**Comment by antonvh** [comment]
It is integrated now
